File: ruby/sorting/quicksort.rb
def quicksort(arr)
  if arr.length <= 1 
    return arr
  end
  midpoint = arr.length.div(2)
  pivot = arr[midpoint]
  left = []
  right = []
  equal = []
  arr.each do |num|
    if num < pivot
      left << num
    elsif num > pivot
      right << num
    else
      equal << num
    end
  end
  return quicksort(left) + equal + quicksort(right)
end

print quicksort([5, 1, 4, 2, 3])

File: python/sorting/quick_sort.py
def quicksort_middle_pivot_asc(arr):
  if len(arr) <= 1:
    return arr
  
  pivot = arr[len(arr) // 2]
  left = [x for x in arr if x < pivot]
  middle = [x for x in arr if x == pivot]
  right = [x for x in arr if x > pivot]
  return quicksort_middle_pivot_asc(left) + middle + quicksort_middle_pivot_asc(right)

def quicksort_middle_pivot_desc(arr):
  if len(arr) <= 1:
    return arr
  pivot = arr[len(arr) // 2]
  left = [x for x in arr if x < pivot]
  middle = [x for x in arr if x == pivot]
  right = [x for x in arr if x > pivot]
  return quicksort_middle_pivot_desc(right) + middle + quicksort_middle_pivot_desc(left)



arr = [5, 3, 1, 2, 4]
print(quicksort_middle_pivot_asc(arr))
print(quicksort_middle_pivot_desc(arr))

  
  